File: crates/zk-sync/src/diff3.rs
//! Line-based three-way text merge (diff3) for Markdown note bodies (ZK-052).
//!
//! In accordance with MASTER_SPEC.md § 10.3:
//! - Non-overlapping line edits auto-merge cleanly;
//! - Identical concurrent edits merge without conflict;
//! - Overlapping divergent line edits become explicit conflicts with diff3 markers;
//! - No side is silently discarded under any condition;
//! - Deterministic behavior across all platforms.

use std::collections::HashMap;

/// An explicit overlapping line-level conflict between local and remote edits.
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct BodyConflict {
    /// Approximate 1-based line number in the base document where the conflict began.
    pub base_line_start: usize,
    /// Conflicting lines from the base revision.
    pub base_lines: Vec<String>,
    /// Conflicting lines from the local revision.
    pub local_lines: Vec<String>,
    /// Conflicting lines from the remote revision.
    pub remote_lines: Vec<String>,
}

/// The result of a line-based three-way diff merge of text bodies.
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct Diff3Result {
    /// Merged text body. If `is_clean` is true, non-overlapping edits were merged cleanly.
    /// If `is_clean` is false, contains formatted conflict markers preserving all sides.
    pub merged_text: String,
    /// Whether the text body merged cleanly without any overlapping conflicts.
    pub is_clean: bool,
    /// List of explicit overlapping conflicts encountered.
    pub conflicts: Vec<BodyConflict>,
}

/// Computes Longest Common Subsequence (LCS) matched line index pairs between slices `a` and `b`.
///
/// Optimizes common prefix and suffix first, running dynamic programming on the differing core.
fn lcs_matched_pairs(a: &[&str], b: &[&str]) -> Vec<(usize, usize)> {
    if a.is_empty() || b.is_empty() {
        return Vec::new();
    }

    // 1. Trim common prefix
    let mut prefix_len = 0;
    while prefix_len < a.len() && prefix_len < b.len() && a[prefix_len] == b[prefix_len] {
        prefix_len += 1;
    }

    // 2. Trim common suffix
    let mut suffix_len = 0;
    while suffix_len < a.len() - prefix_len
        && suffix_len < b.len() - prefix_len
        && a[a.len() - 1 - suffix_len] == b[b.len() - 1 - suffix_len]
    {
        suffix_len += 1;
    }

    let a_mid = &a[prefix_len..a.len() - suffix_len];
    let b_mid = &b[prefix_len..b.len() - suffix_len];

    let mut mid_pairs = Vec::new();

    if !a_mid.is_empty() && !b_mid.is_empty() {
        let n = a_mid.len();
        let m = b_mid.len();

        if n.saturating_mul(m) <= 4_000_000 {
            let mut dp = vec![vec![0u32; m + 1]; n + 1];
            for i in (0..n).rev() {
                for j in (0..m).rev() {
                    if a_mid[i] == b_mid[j] {
                        dp[i][j] = 1 + dp[i + 1][j + 1];
                    } else {
                        dp[i][j] = dp[i + 1][j].max(dp[i][j + 1]);
                    }
                }
            }

            let mut i = 0;
            let mut j = 0;
            while i < n && j < m {
                if a_mid[i] == b_mid[j] {
                    mid_pairs.push((prefix_len + i, prefix_len + j));
                    i += 1;
                    j += 1;
                } else if dp[i + 1][j] >= dp[i][j + 1] {
                    i += 1;
                } else {
                    j += 1;
                }
            }
        }
    }

    let total_matches = prefix_len + mid_pairs.len() + suffix_len;
    let mut result = Vec::with_capacity(total_matches);

    for k in 0..prefix_len {
        result.push((k, k));
    }

    result.extend(mid_pairs);

    for k in 0..suffix_len {
        result.push((a.len() - suffix_len + k, b.len() - suffix_len + k));
    }

    result
}

/// Performs a deterministic line-based three-way merge (diff3) on text bodies.
///
/// Rules:
/// - Non-overlapping edits auto-merge cleanly.
/// - Identical concurrent edits merge without conflict.
/// - Overlapping divergent edits become explicit conflict blocks demarcated with standard
///   diff3 conflict markers (`<<<<<<< LOCAL`, `||||||| BASE`, `=======`, `>>>>>>> REMOTE`).
/// - No text from any side is silently discarded.
pub fn diff3_merge(base: &str, local: &str, remote: &str) -> Diff3Result {
    // Fast path: if local and remote are identical, return local directly
    if local == remote {
        return Diff3Result {
            merged_text: local.to_string(),
            is_clean: true,
            conflicts: Vec::new(),
        };
    }

    // Fast path: if remote equals base, return local
    if remote == base {
        return Diff3Result {
            merged_text: local.to_string(),
            is_clean: true,
            conflicts: Vec::new(),
        };
    }

    // Fast path: if local equals base, return remote
    if local == base {
        return Diff3Result {
            merged_text: remote.to_string(),
            is_clean: true,
            conflicts: Vec::new(),
        };
    }

    // Normalize newlines to \n to guarantee platform-independent deterministic line splitting
    let base_norm = base.replace("\r\n", "\n").replace('\r', "\n");
    let local_norm = local.replace("\r\n", "\n").replace('\r', "\n");
    let remote_norm = remote.replace("\r\n", "\n").replace('\r', "\n");

    let base_lines: Vec<&str> = if base_norm.is_empty() {
        Vec::new()
    } else {
        base_norm.split('\n').collect()
    };
    let local_lines: Vec<&str> = if local_norm.is_empty() {
        Vec::new()
    } else {
        local_norm.split('\n').collect()
    };
    let remote_lines: Vec<&str> = if remote_norm.is_empty() {
        Vec::new()
    } else {
        remote_norm.split('\n').collect()
    };

    let pairs_bl = lcs_matched_pairs(&base_lines, &local_lines);
    let pairs_br = lcs_matched_pairs(&base_lines, &remote_lines);

    let br_map: HashMap<usize, usize> = pairs_br.into_iter().collect();

    // Common anchors: lines present in base that were matched in both local and remote
    let mut common_anchors: Vec<(usize, usize, usize)> = Vec::new();
    for (b, l) in pairs_bl {
        if let Some(&r) = br_map.get(&b) {
            common_anchors.push((b, l, r));
        }
    }

    let mut output_lines: Vec<String> = Vec::new();
    let mut conflicts: Vec<BodyConflict> = Vec::new();

    let mut prev_b = 0;
    let mut prev_l = 0;
    let mut prev_r = 0;

    let process_chunk = |b_start: usize,
                         b_end: usize,
                         l_start: usize,
                         l_end: usize,
                         r_start: usize,
                         r_end: usize,
                         out: &mut Vec<String>,
                         confs: &mut Vec<BodyConflict>| {
        let chunk_base = &base_lines[b_start..b_end];
        let chunk_local = &local_lines[l_start..l_end];
        let chunk_remote = &remote_lines[r_start..r_end];

        let local_changed = chunk_local != chunk_base;
        let remote_changed = chunk_remote != chunk_base;

        match (local_changed, remote_changed) {
            (false, false) => {
                // Unchanged in both
                for line in chunk_base {
                    out.push((*line).to_string());
                }
            }
            (true, false) => {
                // Changed only locally
                for line in chunk_local {
                    out.push((*line).to_string());
                }
            }
            (false, true) => {
                // Changed only remotely
                for line in chunk_remote {
                    out.push((*line).to_string());
                }
            }
            (true, true) => {
                if chunk_local == chunk_remote {
                    // Identical concurrent change
                    for line in chunk_local {
                        out.push((*line).to_string());
                    }
                } else {
                    // Overlapping divergent conflict
                    confs.push(BodyConflict {
                        base_line_start: b_start + 1,
                        base_lines: chunk_base.iter().map(|s| (*s).to_string()).collect(),
                        local_lines: chunk_local.iter().map(|s| (*s).to_string()).collect(),
                        remote_lines: chunk_remote.iter().map(|s| (*s).to_string()).collect(),
                    });

                    // Format conflict chunk with standard markers
                    out.push("<<<<<<< LOCAL".to_string());
                    for line in chunk_local {
                        out.push((*line).to_string());
                    }
                    out.push("||||||| BASE".to_string());
                    for line in chunk_base {
                        out.push((*line).to_string());
                    }
                    out.push("=======".to_string());
                    for line in chunk_remote {
                        out.push((*line).to_string());
                    }
                    out.push(">>>>>>> REMOTE".to_string());
                }
            }
        }
    };

    for (b, l, r) in common_anchors {
        process_chunk(
            prev_b,
            b,
            prev_l,
            l,
            prev_r,
            r,
            &mut output_lines,
            &mut conflicts,
        );

        // Emit the anchor line itself
        output_lines.push(base_lines[b].to_string());

        prev_b = b + 1;
        prev_l = l + 1;
        prev_r = r + 1;
    }

    // Process final region after last anchor
    process_chunk(
        prev_b,
        base_lines.len(),
        prev_l,
        local_lines.len(),
        prev_r,
        remote_lines.len(),
        &mut output_lines,
        &mut conflicts,
    );

    let merged_text = output_lines.join("\n");
    let is_clean = conflicts.is_empty();

    Diff3Result {
        merged_text,
        is_clean,
        conflicts,
    }
}

#[cfg(test)]
#[allow(clippy::expect_used, clippy::unwrap_used, clippy::panic)]
mod tests {
    use super::*;

    #[test]
    fn test_diff3_all_identical() {
        let text = "Line 1\nLine 2\nLine 3";
        let res = diff3_merge(text, text, text);
        assert!(res.is_clean);
        assert_eq!(res.merged_text, text);
        assert!(res.conflicts.is_empty());
    }

    #[test]
    fn test_diff3_non_overlapping_edits_auto_merge() {
        let base = "Section 1\nIntro\n\nSection 2\nMiddle\n\nSection 3\nOutro";
        let local = "Section 1\nIntro modified by local\n\nSection 2\nMiddle\n\nSection 3\nOutro";
        let remote = "Section 1\nIntro\n\nSection 2\nMiddle\n\nSection 3\nOutro modified by remote";

        let res = diff3_merge(base, local, remote);
        assert!(res.is_clean);
        assert!(res.conflicts.is_empty());

        let expected = "Section 1\nIntro modified by local\n\nSection 2\nMiddle\n\nSection 3\nOutro modified by remote";
        assert_eq!(res.merged_text, expected);
    }

    #[test]
    fn test_diff3_identical_concurrent_edits() {
        let base = "Line 1\nLine 2\nLine 3";
        let local = "Line 1\nLine 2 changed by both\nLine 3";
        let remote = "Line 1\nLine 2 changed by both\nLine 3";

        let res = diff3_merge(base, local, remote);
        assert!(res.is_clean);
        assert_eq!(res.merged_text, local);
    }

    #[test]
    fn test_diff3_overlapping_edits_become_explicit_conflict() {
        let base = "Line 1\nOriginal line 2\nLine 3";
        let local = "Line 1\nAlice line 2\nLine 3";
        let remote = "Line 1\nBob line 2\nLine 3";

        let res = diff3_merge(base, local, remote);
        assert!(!res.is_clean);
        assert_eq!(res.conflicts.len(), 1);

        let conflict = &res.conflicts[0];
        assert_eq!(conflict.base_lines, vec!["Original line 2"]);
        assert_eq!(conflict.local_lines, vec!["Alice line 2"]);
        assert_eq!(conflict.remote_lines, vec!["Bob line 2"]);

        // Verify conflict markers are included and no side is silently discarded
        assert!(res.merged_text.contains("<<<<<<< LOCAL"));
        assert!(res.merged_text.contains("Alice line 2"));
        assert!(res.merged_text.contains("||||||| BASE"));
        assert!(res.merged_text.contains("Original line 2"));
        assert!(res.merged_text.contains("======="));
        assert!(res.merged_text.contains("Bob line 2"));
        assert!(res.merged_text.contains(">>>>>>> REMOTE"));
    }

    #[test]
    fn test_diff3_additions_at_both_ends() {
        let base = "Middle Line";
        let local = "Top Line Added\nMiddle Line";
        let remote = "Middle Line\nBottom Line Added";

        let res = diff3_merge(base, local, remote);
        assert!(res.is_clean);
        assert_eq!(
            res.merged_text,
            "Top Line Added\nMiddle Line\nBottom Line Added"
        );
    }

    #[test]
    fn test_diff3_deletion_on_one_side() {
        let base = "Line 1\nLine 2 to delete\nLine 3";
        let local = "Line 1\nLine 3";
        let remote = "Line 1\nLine 2 to delete\nLine 3";

        let res = diff3_merge(base, local, remote);
        assert!(res.is_clean);
        assert_eq!(res.merged_text, "Line 1\nLine 3");
    }
}

File: apps/web/src/utils/diff3.ts
/**
 * Three-Way Text Diff3 and Structured Note Merge Engine (ZK-068).
 *
 * Implements deterministic three-way merge semantics according to MASTER_SPEC.md § 10.2 & 10.3:
 * - Non-overlapping line edits auto-merge cleanly.
 * - Identical concurrent edits merge without conflict.
 * - Overlapping divergent line edits become explicit conflicts with standard diff3 markers:
 *   <<<<<<< LOCAL
 *   =======
 *   >>>>>>> REMOTE
 * - Deterministic set-based tag merging (sorted, deduplicated, lowercase).
 * - Preserves all sides completely without silent data loss.
 */

import { PlaintextNoteDto } from "../worker/protocol.js";

export interface BodyConflict {
  baseLineStart: number;
  baseLines: string[];
  localLines: string[];
  remoteLines: string[];
}

export interface Diff3Result {
  mergedText: string;
  isClean: boolean;
  conflicts: BodyConflict[];
}

/**
 * Computes Longest Common Subsequence (LCS) matched line index pairs between slices a and b.
 * Uses prefix/suffix trimming before DP table computation.
 */
export function lcsMatchedPairs(a: string[], b: string[]): Array<[number, number]> {
  if (a.length === 0 || b.length === 0) {
    return [];
  }

  // 1. Trim common prefix
  let prefixLen = 0;
  while (prefixLen < a.length && prefixLen < b.length && a[prefixLen] === b[prefixLen]) {
    prefixLen++;
  }

  // 2. Trim common suffix
  let suffixLen = 0;
  while (
    suffixLen < a.length - prefixLen &&
    suffixLen < b.length - prefixLen &&
    a[a.length - 1 - suffixLen] === b[b.length - 1 - suffixLen]
  ) {
    suffixLen++;
  }

  const aMid = a.slice(prefixLen, a.length - suffixLen);
  const bMid = b.slice(prefixLen, b.length - suffixLen);

  const midPairs: Array<[number, number]> = [];

  if (aMid.length > 0 && bMid.length > 0) {
    const n = aMid.length;
    const m = bMid.length;

    if (n * m <= 4000000) {
      const stride = m + 1;
      const dp = new Int32Array((n + 1) * stride);

      for (let i = n - 1; i >= 0; i--) {
        for (let j = m - 1; j >= 0; j--) {
          if (aMid[i] === bMid[j]) {
            dp[i * stride + j] = 1 + (dp[(i + 1) * stride + (j + 1)] ?? 0);
          } else {
            const down = dp[(i + 1) * stride + j] ?? 0;
            const right = dp[i * stride + (j + 1)] ?? 0;
            dp[i * stride + j] = Math.max(down, right);
          }
        }
      }

      let i = 0;
      let j = 0;
      while (i < n && j < m) {
        if (aMid[i] === bMid[j]) {
          midPairs.push([prefixLen + i, prefixLen + j]);
          i++;
          j++;
        } else {
          const down = dp[(i + 1) * stride + j] ?? 0;
          const right = dp[i * stride + (j + 1)] ?? 0;
          if (down >= right) {
            i++;
          } else {
            j++;
          }
        }
      }
    }
  }

  const result: Array<[number, number]> = [];
  for (let k = 0; k < prefixLen; k++) {
    result.push([k, k]);
  }
  for (const p of midPairs) {
    result.push(p);
  }
  for (let k = 0; k < suffixLen; k++) {
    result.push([a.length - suffixLen + k, b.length - suffixLen + k]);
  }

  return result;
}

/**
 * Line-based three-way text merge (diff3).
 */
export function diff3Merge(base: string, local: string, remote: string): Diff3Result {
  // Fast path: if local equals remote, both made the exact same change
  if (local === remote) {
    return {
      mergedText: local,
      isClean: true,
      conflicts: [],
    };
  }

  // Fast path: if remote equals base, keep local
  if (remote === base) {
    return {
      mergedText: local,
      isClean: true,
      conflicts: [],
    };
  }

  // Fast path: if local equals base, keep remote
  if (local === base) {
    return {
      mergedText: remote,
      isClean: true,
      conflicts: [],
    };
  }

  // Normalize newlines to \n
  const baseNorm = base.replace(/\r\n/g, "\n").replace(/\r/g, "\n");
  const localNorm = local.replace(/\r\n/g, "\n").replace(/\r/g, "\n");
  const remoteNorm = remote.replace(/\r\n/g, "\n").replace(/\r/g, "\n");

  const baseLines = baseNorm.length === 0 ? [] : baseNorm.split("\n");
  const localLines = localNorm.length === 0 ? [] : localNorm.split("\n");
  const remoteLines = remoteNorm.length === 0 ? [] : remoteNorm.split("\n");

  const pairsBL = lcsMatchedPairs(baseLines, localLines);
  const pairsBR = lcsMatchedPairs(baseLines, remoteLines);

  const brMap = new Map<number, number>(pairsBR);

  // Common anchors: lines present in base that were matched in both local and remote
  const commonAnchors: Array<[number, number, number]> = [];
  for (const [b, l] of pairsBL) {
    const r = brMap.get(b);
    if (r !== undefined) {
      commonAnchors.push([b, l, r]);
    }
  }

  const outputLines: string[] = [];
  const conflicts: BodyConflict[] = [];

  let prevB = 0;
  let prevL = 0;
  let prevR = 0;

  const processChunk = (
    bStart: number,
    bEnd: number,
    lStart: number,
    lEnd: number,
    rStart: number,
    rEnd: number
  ) => {
    const chunkBase = baseLines.slice(bStart, bEnd);
    const chunkLocal = localLines.slice(lStart, lEnd);
    const chunkRemote = remoteLines.slice(rStart, rEnd);

    const localEqualBase =
      chunkLocal.length === chunkBase.length &&
      chunkLocal.every((val, idx) => val === chunkBase[idx]);

    const remoteEqualBase =
      chunkRemote.length === chunkBase.length &&
      chunkRemote.every((val, idx) => val === chunkBase[idx]);

    if (localEqualBase && remoteEqualBase) {
      // Unchanged in both
      for (const line of chunkBase) {
        outputLines.push(line);
      }
    } else if (!localEqualBase && remoteEqualBase) {
      // Changed only locally
      for (const line of chunkLocal) {
        outputLines.push(line);
      }
    } else if (localEqualBase && !remoteEqualBase) {
      // Changed only remotely
      for (const line of chunkRemote) {
        outputLines.push(line);
      }
    } else {
      // Both changed
      const localEqualRemote =
        chunkLocal.length === chunkRemote.length &&
        chunkLocal.every((val, idx) => val === chunkRemote[idx]);

      if (localEqualRemote) {
        // Identical concurrent edit
        for (const line of chunkLocal) {
          outputLines.push(line);
        }
      } else {
        // Overlapping divergent conflict
        conflicts.push({
          baseLineStart: bStart + 1,
          baseLines: chunkBase,
          localLines: chunkLocal,
          remoteLines: chunkRemote,
        });

        outputLines.push("<<<<<<< LOCAL");
        for (const line of chunkLocal) {
          outputLines.push(line);
        }
        outputLines.push("=======");
        for (const line of chunkRemote) {
          outputLines.push(line);
        }
        outputLines.push(">>>>>>> REMOTE");
      }
    }
  };

  for (const [b, l, r] of commonAnchors) {
    processChunk(prevB, b, prevL, l, prevR, r);

    // Emit the anchor line itself
    const anchorLine = baseLines[b];
    if (anchorLine !== undefined) {
      outputLines.push(anchorLine);
    }

    prevB = b + 1;
    prevL = l + 1;
    prevR = r + 1;
  }

  // Process final region after last anchor
  processChunk(prevB, baseLines.length, prevL, localLines.length, prevR, remoteLines.length);

  const mergedText = outputLines.join("\n");
  const isClean = conflicts.length === 0;

  return {
    mergedText,
    isClean,
    conflicts,
  };
}

/**
 * Three-way tag set merge.
 * - Tags in base are kept unless deleted by local or remote.
 * - Tags not in base are added if added by local or remote.
 * - Canonical output: lowercase, trimmed, deduplicated, sorted.
 */
export function mergeTags(
  baseTags: string[],
  localTags: string[],
  remoteTags: string[]
): string[] {
  const baseSet = new Set(baseTags.map((t) => t.trim().toLowerCase()).filter(Boolean));
  const localSet = new Set(localTags.map((t) => t.trim().toLowerCase()).filter(Boolean));
  const remoteSet = new Set(remoteTags.map((t) => t.trim().toLowerCase()).filter(Boolean));

  const allTags = new Set<string>([...baseSet, ...localSet, ...remoteSet]);
  const merged: string[] = [];

  for (const tag of allTags) {
    const inBase = baseSet.has(tag);
    const inLocal = localSet.has(tag);
    const inRemote = remoteSet.has(tag);

    if (inBase) {
      // Kept unless removed by either side
      if (inLocal && inRemote) {
        merged.push(tag);
      }
    } else {
      // Added if added by either side
      if (inLocal || inRemote) {
        merged.push(tag);
      }
    }
  }

  merged.sort((a, b) => a.localeCompare(b));
  return merged;
}

export interface ThreeWayNoteMergeResult {
  candidate: PlaintextNoteDto;
  isClean: boolean;
  hasTitleConflict: boolean;
  hasBodyConflict: boolean;
  diffResult: Diff3Result;
}

/**
 * Executes a full 3-way structured merge between BASE, LOCAL, and REMOTE notes.
 */
export function threeWayMergeNotes(
  baseNote: PlaintextNoteDto | null,
  localNote: PlaintextNoteDto,
  remoteNote: PlaintextNoteDto
): ThreeWayNoteMergeResult {
  const baseTitle = baseNote?.title || "";
  const baseBody = baseNote?.body || "";
  const baseTags = baseNote?.tags || [];

  // 1. Title merge
  let mergedTitle = localNote.title;
  let hasTitleConflict = false;

  const localTitleChanged = localNote.title !== baseTitle;
  const remoteTitleChanged = remoteNote.title !== baseTitle;

  if (!localTitleChanged && !remoteTitleChanged) {
    mergedTitle = baseTitle;
  } else if (localTitleChanged && !remoteTitleChanged) {
    mergedTitle = localNote.title;
  } else if (!localTitleChanged && remoteTitleChanged) {
    mergedTitle = remoteNote.title;
  } else {
    // Both changed
    if (localNote.title === remoteNote.title) {
      mergedTitle = localNote.title;
    } else {
      hasTitleConflict = true;
      mergedTitle = localNote.title; // Default candidate to local title
    }
  }

  // 2. Body merge (diff3)
  const diffResult = diff3Merge(baseBody, localNote.body, remoteNote.body);

  // 3. Tags merge
  const mergedTags = mergeTags(baseTags, localNote.tags, remoteNote.tags);

  // 4. Attachments merge (3-way set merge)
  const baseAttachments = baseNote?.attachments || [];
  const localAttachments = localNote.attachments || [];
  const remoteAttachments = remoteNote.attachments || [];
  const mergedAttachments = mergeTags(baseAttachments, localAttachments, remoteAttachments);

  const isClean = !hasTitleConflict && diffResult.isClean;

  const candidate: PlaintextNoteDto = {
    id: localNote.id,
    title: mergedTitle,
    body: diffResult.mergedText,
    tags: mergedTags,
    attachments: mergedAttachments,
    createdAt: localNote.createdAt || remoteNote.createdAt || new Date().toISOString(),
    updatedAt: new Date().toISOString(),
  };

  return {
    candidate,
    isClean,
    hasTitleConflict,
    hasBodyConflict: !diffResult.isClean,
    diffResult,
  };
}
